src/thor/costmatrix.cc

```diff
@@ -734,10 +734,17 @@ void CostMatrix::SetSources(GraphReader& graphreader,
                                                                   &source_edgelabel_[index]));
     source_hierarchy_limits_[index] = costing_->GetHierarchyLimits();
 
+    // Only skip inbound edges if we have other options
+    bool has_other_edges = false;
+    std::for_each(origin.correlation().edges().begin(), origin.correlation().edges().end(),
+                  [&has_other_edges](const valhalla::PathEdge& e) {
+                    has_other_edges = has_other_edges || !e.end_node();
+                  });
+
     // Iterate through edges and add to adjacency list
     for (const auto& edge : origin.correlation().edges()) {
       // If origin is at a node - skip any inbound edge (dist = 1)
-      if (edge.end_node()) {
+      if (has_other_edges && edge.end_node()) {
```

Comment:
Ah yeah, good point! In fact that is also missing from the `TimeDistanceMatrix` when setting the destinations and the origin for many:one. I added it in #3773 .